### PR TITLE
feature: add a `diff` command to be able to retrieve a doc change

### PR DESCRIPTION
**Pull request by paulRbr**
This is a new command `bump diff` which allow to fetch a documentation
change between an existing documentation and a provided definition
FILE.

```
Get a comparaison diff with your documentation from the given file or URL

USAGE
  $ bump diff FILE

ARGUMENTS
  FILE  Path or URL to your API documentation file. OpenAPI (2.0 to 3.1.0) and AsyncAPI (2.0)
        specifications are currently supported.

OPTIONS
  -b, --hub=hub      Hub id or slug. Can be provided via BUMP_HUB_ID environment variable

  -d, --doc=doc      (required) Documentation public id or slug. Can be provided via BUMP_ID environment
                     variable

  -h, --help         show CLI help

  -t, --token=token  (required) Documentation or Hub token. Can be provided via BUMP_TOKEN environment
                     variable

EXAMPLES
  Compare a potential new version with an existing documentation.

  $ bump diff FILE --doc <your_doc_id_or_slug> --token <your_doc_token>
  * Let's compare the given definition version... done
  Removed: GET /compare
  Added: GET /versions/{versionId}/changes

  Compare a non modified definition FILE with an existing documentation

  $ bump diff FILE --doc <doc_slug> --token <your_doc_token>
  * Let's compare the given definition version... done
    ›   Warning: Your documentation has not changed!


```

For now it uses polling mechanism (a http request every second for max
30 seconds) to fetch the diff result. We should later expose a
websocket public API so the CLI can simply listen to events coming
from the WS for deploy status.

## Demo

**`git diff`** on our openapi.yml spec
```
> git diff
diff --git a/doc/api/v1/openapi.v3.yml b/doc/api/v1/openapi.v3.yml
index af8a1d743..c311d7808 100644
--- a/doc/api/v1/openapi.v3.yml
+++ b/doc/api/v1/openapi.v3.yml
@@ -216,10 +216,6 @@ components:
           type: string
           description: "Unique id of your newly deployed version."
           example: 2361df99-3467-4c80-a0cc-45c9fe565812
-        doc_id:
-          type: string
-          description: "Unique id of your documentation."
-          example: 3ef8f52f-9056-4113-840e-2f7183b90e06
         doc_public_url:
           type: string
           description: "The public URL of your documentation."
```

**`bump diff`**
```
> time BUMP_TOKEN=123abc BUMP_HOST=http://localhost:3000/ ./bin/run diff -d bump --hub yoyo ../bump/doc/api/v1/openapi.v3.yml                    
* Let's compare the given definition version... done
Updated: POST /versions
  Response modified: 201
    Body attribute removed: doc_id

real    0m5,487s
user    0m1,505s
sys     0m0,104s
```

### CODE REVIEW

File: src/commands/diff.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  $ bump diff FILE --doc <doc_slug> --token <your_doc_token>
+  * Let's compare the given definition version... done
+   ›   Warning: Your documentation has not changed!
+`,
+  ];
+
+  static flags = {
+    help: flags.help({ char: 'h' }),
+    doc: flags.doc(),
+    hub: flags.hub(),
```
[the remark this answers is not in the document]
Comment (paulRbr) [reply src/commands/diff.ts:41]: For the same reason we need it when we deploy a new version: a doc slug is not unique. We can have many docs with the same slug name but under different hubs.

[the remark this answers is not in the document]
Comment (Polo2) [reply src/commands/diff.ts:41]: OK, thanks for your answer.

I've been confused by 'Documentation public id _or_ slug'. Indeed, public_id is uniq but not slug

File: package.json [hunk 2]
```diff
@@ -66,6 +66,7 @@
     "fmt": "eslint . --ext .ts --config .eslintrc --fix",
     "pack": "oclif-dev pack",
     "postpack": "rm -f oclif.manifest.json",
+    "postinstall": "npm run build",
```
Comment (Polo2) [inline package.json:69]: 🙈  Sorry to mention it, but since every keys are sorted alphabetically, maybe 'postinstall' should arrive before 'postpack' ?

**Review by Polo2 — APPROVED** [review]
Great job 👏  that's a very nice feature introduced by this PR 🎉  

I'm definitely not skilled enough to give you a complete review, but new code looks great, form my position I can approve this PR

File: src/commands/diff.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      case 204:
+        const version: VersionResponse = diffResponse.data;
+
+        this.d(`Received version:`);
+        this.d(version);
+        return version;
+        break;
+      case 202:
+        this.d('Waiting 1 sec before next pool');
+        await this.pollingDelay();
```
Comment (scharrier) [inline src/commands/diff.ts:121]: Should we add a loader for the user, here?

```
* Let's compare the given definition version.................... done
```

**Review by scharrier — COMMENTED** [review]
I left some comments, even if I know that this whole PR will need an update according to our latest discussions/choices.

This will be so nice to use! 👏

File: src/api/index.ts [hunk 4]
```diff
@@ -7,6 +7,7 @@ import {
   PreviewResponse,
   VersionRequest,
   VersionResponse,
+  ChangesResponse,
```
[the remark this answers is not in the document]
Comment (paulRbr) [reply src/api/index.ts]: :thinking: I don't know how you managed to comment on an old commit. I removed this `ChangesResponse` model last night :thinking:

File: src/commands/diff.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+import { fileArg } from '../args';
+import { cli } from '../cli';
+import { VersionRequest, VersionResponse, ChangesResponse } from '../api/models';
+
+export default class Diff extends Command {
+  static description =
+    'Get a comparaison diff with your documentation from the given file or URL';
+
+  static examples = [
+    `Compare a potential new version with an existing documentation:
```
[the remark this answers is not in the document]
Comment (paulRbr) [reply src/commands/diff.ts:14]: :+1:

File: src/commands/diff.ts (hunk 3 above)
Comment (paulRbr) [reply src/commands/diff.ts:121]: Not sure to understand your question, there's already a loader present:

Peek 24-06-2021 17-40

File: src/commands/diff.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+          version.diff_lines.forEach((line) => cli.log(line));
+        } else {
+          this.warn('There were no structural changes in your new definition');
+        }
+
+        return version;
+        break;
+      case 204:
+        cli.action.stop();
+        this.warn('Your documentation has not changed!');
```
[the remark this answers is not in the document]
Comment (paulRbr) [reply src/commands/diff.ts:90]: I'll remove the bang in all warning message. It's a bit to “enthusiastic” :sweat_smile:

File: package.json (hunk 2 above)
Comment (paulRbr) [reply package.json]: I've removed the commit. It's not necessary, it's was here for local debugging only :see_no_evil:

File: src/commands/diff.ts (hunk 3 above)
Comment (scharrier) [reply src/commands/diff.ts:120]: Oops, sorry, I didn't see it. Thanks for the explanation!

**Review by scharrier — APPROVED** [review]
LGTM! The final changes you did really simplify the whole thing! 👏

**Pull request merged**